inference-engine/src/gna_plugin/transformations/decompose_2d_conv.cpp

```diff
@@ -0,0 +1,597 @@
+    std::shared_ptr<ngraph::Node>last_op_in_sequence_for_replacement;
+    std::shared_ptr<ngraph::Node>bias_const;
+    size_t conv_count;
+    size_t pool_size_width;
+    size_t pool_stride_width;
+    // TODO: currently 2D max pool is not supported
+    //size_t pool_size_height;
+    //size_t pool_stride_height;
+};
+
+static bool VerifyAndGetConvData(std::shared_ptr<ngraph::opset7::Convolution> conv, ConvData& conv_data) {
+    const auto& input = conv->input_value(0);
+    const auto& filters = conv->input_value(1);
+
+    // We support only batch == 1
+    if (input.get_shape()[0] != 1) {
+        return false;
+    }
+
+    size_t filter_height = filters.get_shape()[2];
+    size_t filter_width = filters.get_shape()[3];
+
+    if (filter_width > GNALimitations::copyMaxGrouping || filter_height > GNALimitations::copyMaxGrouping) {
+        return false;
+    }
+
+    GetConvData(conv, conv_data);
+
+    IE_ASSERT(conv_data.output_channel_count == conv->get_output_shape(0)[1]);
+
+    return true;
+}
+
+template<typename T>
+static std::shared_ptr<ngraph::Node> VerifyBiasAndCreateConst(std::shared_ptr<ngraph::opset7::Add> conv_bias, const ConvData& conv_data) {
+    auto add_const = std::dynamic_pointer_cast<ngraph::op::Constant>(conv_bias->input_value(1).get_node_shared_ptr());
+
+    if (add_const) {
+        auto bias_size = shape_size(add_const->get_shape());
+
+        // The add may be a normal add not conv bias, then we just go further
+        if (bias_size == conv_data.filter_count) {
+            const auto* srd_data_pointer = add_const->get_data_ptr<T>();
+            std::vector<T> bias_values(srd_data_pointer, srd_data_pointer + bias_size);
+            return ngraph::opset7::Constant::create(conv_data.element_type, ngraph::Shape{1, bias_size , 1, 1}, bias_values);
+        }
+    }
+    // Bias size does not match (or dynamic bias), can't decompose such convolution
+    return nullptr;
+}
+
+static bool VerifyMaxPool(GraphData& graph_data, std::shared_ptr<ngraph::opset7::MaxPool> max_pool) {
+    auto pool_filter = max_pool->get_kernel();
+    auto pool_strides = max_pool->get_strides();
+
+    // Check Max Pool padding and limitations
+    if (max_pool->get_auto_pad() != ngraph::op::PadType::VALID ||
+        pool_filter.size() != 2 || pool_strides.size() != 2 ||
+        pool_filter[0] > GNALimitations::maxPoolMaxWindowSize)
+        return false;
+
+    graph_data.pool_size_width = pool_filter[1];
+    graph_data.pool_stride_width = pool_strides[1];
+    return true;
+}
+
+static bool ShouldDecompose(GraphData& graph_data, const ConvData& conv_data) {
+    // Check if split of plane due to GNA HW limitations of 768 filter elements is possible
+    graph_data.conv_count = 1;
+    size_t total_factorized_conv_channel_count = (conv_data.input_channel_count * conv_data.filter_height * conv_data.filter_width);
+    while (total_factorized_conv_channel_count / graph_data.conv_count > GNALimitations::convFilterMaxSize ||
```

Comment:
I think it is better to calculate  `conv_count` in a separate function. It is pretty unexpected for me that the function which checks the possibility of decomposition fills some data.

Reply:
@sirzabek Currently it is not true due to the pooling check. Otherwise, this function can be easily renamed to CalculateConvCout and declare it void instead of bool, because conv_count gives the answer.

Reply:
I don't see it as a problem that pooling limitation is checked here, in a function called "should" and not "can", because we're able to check all the limitations at this stage only anyway. But in fact in my branch where I'm working on FQ support I've already moved it to an earlier stage, when other limitations are checked, so I think we agree that there is a better place for it.

But conv_count doesn't give the whole answer - we're still checking for dilations and 2D convolutions here. _Any_ of these three conditions influences the result, hence the bool.

So my suggestion is then to move pooling check from here and leave the rest as it is now.

Reply:
To make it shorter I created a separate function CalculateConvCount.